### PR TITLE
MRG: Add return type hints to `read_evokeds()`, `read_raw()`

### PR DESCRIPTION
**Pull request by hoechenberger**
Closes #12249
x-ref #12243

This PR adds return type hints for `read_evokeds()` and `read_raw()`.

Screenshot 2023-11-30 at 18 43 30

Screenshot 2023-11-30 at 18 44 18

I decided to annotate `read_raw()` as `BaseRaw`, as this is the parent class shared among all raw readers.

**Neat:**
Turns out that since all **specific** raw readers (`read_raw_*()`) are structured in a simple fashion and have `return RawXXX()` as their last line of code, VS Code can trivially find out the correct return types of those functions without any additional type hints!

Same goes for `read_epochs()`, hence no changes are required there either.

Still, our docstrings are, strictly speaking, incorrect; `read_raw_X()` returns a `RawX` instance, not `Raw`; and `read_epochs()` returns `EpochsFIF`, not `Epochs`. Just something to keep in mind for a future cleanup or so.

### CODE REVIEW

**Comment by hoechenberger** [comment]
CI errors seem unrelated 🤔

**Comment by larsoner** [comment]
Looks good to me so far. Is it reasonable to add `mypy` to check just these functions, or no? If so I can try adding it to this PR. And @hoechenberger we probably need a `doc/changes/devel.rst` update mentioning that some return type hints have been added.

**Comment by hoechenberger** [comment]
I will look into this, @larsoner

**Comment by hoechenberger** [comment]
I will write a changeog entry later

**Comment by larsoner** [comment]
> I don't have time to adjust the CI configurations etc...  I've gotten red squiggly lines and no tab completion with BunchConstNamed ever since I started using VS Code, so something is not quite right there. Not sure how to fix it properly. I wonder if we should simply switch to SimpleNamespace

Maybe but for now I just ignore `attr-defined`, we hit it on other stuff, too, so it'll be a bigger undertaking. Now added to CIs via pre-commit.ci and it's passing, as is `pre-commit run --all` locally and `mypy mne/` locally.

Based on the link @drammock posted in #12249, I think we're on step https://mypy.readthedocs.io/en/stable/existing_code.html#prioritise-annotating-widely-imported-modules at this point, so @hoechenberger if you add a `devel.rst` update we should be good to go here I think.

File: pyproject.toml [hunk 1]
```diff
@@ -313,3 +314,32 @@ ignore_directives = [
...
 ignore_messages = "^.*(Unknown target name|Undefined substitution referenced)[^`]*$"
+
+[tool.mypy]
+ignore_errors = true
+scripts_are_modules = true
+
+[[tool.mypy.overrides]]
+module = ['mne.evoked', 'mne.io']
+ignore_errors = false
+# Ignore "attr-defined" until we fix stuff like:
```
Comment (larsoner) [inline pyproject.toml:325]: @hoechenberger feel free to look here at what I had to add, basically it gives us a todo list to tackle at some point when people are motivated (or things to live with)

File: pyproject.toml [hunk 2]
```diff
@@ -313,3 +314,32 @@ ignore_directives = [
...
+# - 'Incompatible types in assignment (expression has type "tuple[str, ...]", variable has type "str")'
+# Ignore "operator" until we fix stuff like:
+# - Unsupported operand types for - ("None" and "int")
+disable_error_code = [
+    'attr-defined',
+    'no-untyped-call',
+    'no-untyped-def',
+    'misc',
+    'assignment',
+    'operator',
```
Comment (hoechenberger) [inline pyproject.toml:344]: @larsoner I don't need to add a single one of these to successfully run `mypy mne/` without errors. Did you only add those in anticipation of future errors? I'd rather remove or comment out anything we don't need right now.

Comment (hoechenberger) [reply pyproject.toml:344]: Ah, you were running it with `--strict`. Why?

I don't think we should run with `--strict` for now.

Comment (larsoner) [reply pyproject.toml:344]: To me it's better to know and be explicit here but we can remove it if you want

Comment (larsoner) [reply pyproject.toml:344]: (I'd like to keep the comments though because they do indicate things we can work on in the future)

**Comment by larsoner** [comment]
@hoechenberger fixed a conflict and added to `.git-blame-revs-ignore` from the other PR, good to go from your end?

**Review by larsoner — APPROVED** [review]
@drammock feel free to review and approve then we can merge I think assuming @hoechenberger is also happy

**Comment by hoechenberger** [comment]
@larsoner I think after a while I realized that what bugged me most was an inconsistency between what we were doing in the pre-commit hooks and what happened when I simply ran `mypy ./mne` from the command line.

I've resolved this now in [dc770ec](https://github.com/mne-tools/mne-python/pull/12250/commits/dc770ec85febe6c30780ca7d77ab6610573a09b9)

**Review by drammock — APPROVED** [review]
changes look reasonable / correct.  Doesn't actually work for me in vscode. If I do

```py
import mne

raw = mne.io.read_raw("foo")
raw.<TAB>
```
I get no offers of completion. Since it works for others I'm assuming it's my settings being wrong, but I don't have time to futz with them right now so I'll approve.

**Comment by hoechenberger** [comment]
@drammock Have you tried `Ctrl-Space` instead of `Tab`?

**Comment by drammock** [comment]
I do see the type:

type-hint

but "loading" stays there for tens of seconds. Ctrl+space after `raw.` pops up a bigger box with "loading..." in a bigger font.

**Comment by hoechenberger** [comment]
This looks like a problem with Pylance. Is there anything useful logged in the "Python language server" log window?

**Comment by drammock** [comment]
nothing useful to my eye:

```
2023-12-05 15:08:06.092 [info] [Info  - 3:08:06 PM] (2507452) Pylance language server 2023.11.10 (pyright 088ebaa5) starting
2023-12-05 15:08:06.092 [info] [Info  - 3:08:06 PM] (2507452) Server root directory: /home/drmccloy/.vscode/extensions/ms-python.vscode-pylance-2023.11.10/dist
2023-12-05 15:08:06.104 [info] [Info  - 3:08:06 PM] (2507452) Starting service instance "python"
2023-12-05 15:08:06.143 [info] [Info  - 3:08:06 PM] (2507452) Setting pythonPath for service "python": "/opt/mambaforge/envs/mnedev/bin/python"
2023-12-05 15:08:06.144 [info] [Info  - 3:08:06 PM] (2507452) Setting environmentName for service "python": "3.11.6 (mnedev conda)"
2023-12-05 15:08:06.147 [info] [Info  - 3:08:06 PM] (2507452) Loading pyproject.toml file at /opt/mne/python/pyproject.toml
2023-12-05 15:08:06.163 [info] [Info  - 3:08:06 PM] (2507452) Pyproject file "/opt/mne/python/pyproject.toml" has no "[tool.pyright]" section.
2023-12-05 15:08:06.214 [info] [Info  - 3:08:06 PM] (2507452) Assuming Python version 3.11
2023-12-05 15:08:06.620 [info] [Info  - 3:08:06 PM] (2507452) Found 1845 source files
2023-12-05 15:08:17.420 [info] [Info  - 3:08:17 PM] (2507452) [IDX(2)] Long operation: index execution environment /opt/mne/python (6534ms)
2023-12-05 15:08:17.421 [info] [Info  - 3:08:17 PM] (2507452) [IDX(2)] Long operation: index packages /opt/mne/python (6561ms)
```
all of this was up within ~15 seconds of opening vscode. did a Ctrl+space after `raw.` and waited about 90 seconds, no new log entries.

**Comment by drammock** [comment]
- indexing was True
- indexing limit was at 2000 (MNE has ~1800 files, so shouldn't have been a bottleneck?).  I've changed it to -1
- setting loglevel to Trace made the output pane explode with messages; I don't know where to start (dozens of lines per second, nearly continuously, too fast to read them)

I've also just enabled the "persist indices to disk for all 3rd party libraries" so maybe once it's gone through everything once it will actually become performant? 🤞🏻

**Comment by drammock** [comment]
it finally slowed down. every single message is `info` level. I searched for `io/base` and found these lines:

```
2023-12-05 15:39:08.324 [info] (2603285) [BG(1)] indexing: /opt/mne/python/mne/io/base.py [found 21] (7ms)

...

2023-12-05 15:39:12.858 [info] (2603285) [FG] parsing: /opt/mne/python/mne/io/base.py [fs read 1ms] (60ms)
2023-12-05 15:39:12.872 [info] (2603285) [FG] binding: /opt/mne/python/mne/io/base.py (14ms)
```
ctrl+space still shows "loading..."

**Comment by hoechenberger** [comment]
How did you install MNE?

**Comment by drammock** [comment]
editable pip install.
-------- Original Message --------

**Comment by hoechenberger** [comment]
> FWIW, everything works fine for me, even with an editable install.

If the editable install is working for you out of the box, this indicates that your Pylance is actually **not** picking up the "installed" MNE, but rather the one in your current working directory / workspace. So this is not a valid test for either, editable or non-editable install. You need to use a workspace outside of the `mne-python` directory.

**Comment by cbrnr** [comment]
Can we please switch to Hatchling ASAP? I keep forgetting that editable installs do not work with Setuptools...

And yes, installing it as non-editable in a new venv outside of the MNE source works for me.

**Comment by hoechenberger** [comment]
> OK, so as a dev I will basically never be able to check if these sorts of things are working?

Errm … If you're running all your tests from within the `men-python` directory, you cannot rely on **anything** if your goal is to test behavior in an actual installation. If you run `python -c "import mne"`, Python will pick `mne` from your CWD.

This just to say that this problem affects **all** parts of Python package development. It's not isolated to type hints.

**Comment by cbrnr** [comment]
Yes, that's what CI should test, not any dev.

**Comment by hoechenberger** [comment]
Thanks for testing, @drammock 

 I believe this will be resolved via #12269

Can you try with that branch?

**Comment by drammock** [comment]
> I believe this will be resolved via #12269
> Can you try with that branch?

Here's what I did:

```console
$ cd /opt/mne/python
$ git checkout -b hatchling hoechenberger/hatchling
$ pip install -e .
$ code .
```
No completion, still "Loading..." forever. vscode's bottom bar shows `mnedev` as the active env

If I instead do `code ~/Desktop` then I do get completion, even if vscode's bottom bar shows `mnedev` as the active env.

Is that the expected pattern? Is there a chance that some cruft is causing the failure? in the root of my MNE clone I have (among other things) the folders `.eggs`, `mne.egg-info`, and `.mypy-cache` of which IDK if they might be stale/unnecessary.

**Comment by hoechenberger** [comment]
It's definitely not the expected behavior

Can you try to create a fresh clone of my branch and use that?

**Comment by hoechenberger** [comment]
Do you get completions with other Python packages?

**Comment by larsoner** [comment]
> Can you try to create a fresh clone of my branch and use that?

Or use `git clean -xdf` which will remove all files not under version control. I wonder if it's the `mne.egg-info` specifically, those can be problematic

**Comment by cbrnr** [comment]
This sounds like your VS Code installation is broken. Can you try to reinstall and maybe clean out your user and workspace (.vscode) settings?

**Comment by drammock** [comment]
> Can you try to create a fresh clone of my branch and use that?



> Do you get completions with other Python packages?
yes, e.g., numpy

> Did the problem not occur before we merged this PR?
not sure, I didn't typically use this feature before

> Or use `git clean -xdf`
Nothing changed after doing this.

> This sounds like your VS Code installation is broken. Can you try to reinstall and maybe clean out your user and workspace (.vscode) settings?
`git clean -xdf` nuked the workspace settings. I looked at my global settings, nothing likely to affect this was found

Ultimately I did just fix it, by disabling the `reStructuredText` extension (which runs esbonio language server in the background). Not a big surprise, I've hit trouble with that extension before.

**Comment by drammock** [comment]
rST is not the problem here. it's the vscode extension that relies on silently running sphinx builds in the background in order to offer rST completion hints for crossreferences. A small benefit that is not worth the complexity and cost.

**Comment by hoechenberger** [comment]
I was really just trying to make a joke :) But I do dislike reST, that's for sure :D
Good to know this extension causes problems, I had just recently considered starting to use it; now I probably won't

**Pull request merged**